lib/mongo/server/connection_base.rb

```diff
@@ -213,12 +210,39 @@ def serialize(message, client, buffer = BSON::ByteBuffer.new)
           max_bson_size += MAX_BSON_COMMAND_OVERHEAD
         end
 
-        final_message.serialize(buffer, max_bson_size)
-        if max_message_size &&
-          (buffer.length - start_size) > max_message_size
-        then
+        # RUBY-2234: It is necessary to check that the message size does not
+        # exceed the maximum bson object size before compressing and serializing
+        # the final message.
+        #
+        # This is to avoid the case where the user performs a bulk write
+        # larger than 16MiB which, when compressed, becomes smaller than 16MiB.
+        # If the driver does not split the bulk writes prior to compression,
+        # the entire operation will be sent to the server, which will raise an
+        # error because the uncompressed operation exceeds the maximum bson size.
+        #
+        # To address this problem, we serialize the message prior to compression
+        # and raise an exception if the serialized message exceeds the maximum
+        # bson size.
+
+        # Create a separate buffer that contains the un-compressed message
+        # for the purpose of checking its size. Write any pre-existing contents
+        # from the original buffer into the temporary one.
+        temp_buffer = BSON::ByteBuffer.new
+        temp_buffer.put_bytes(buffer.get_bytes(buffer.length))
```

Comment:
This actually mutates the buffer on the reading side, which we could look into addressing in the future.

Reply:
Could you add a TODO comment noting that the buffer is mutated on this line?